crucible-llvm/src/Lang/Crucible/LLVM/MemModel/MemLog.hs

```diff
@@ -204,47 +205,57 @@ possibleAllocInfo n (MemAllocs xs) = asum (map helper xs)
 
 
 -- | Generalized function for checking whether a memory region ID is allocated.
+--
+-- The first predicate indicates whether the region was allocated, the second
+-- indicates whether it has *not* been freed.
 isAllocatedGeneric ::
   forall sym.
   (IsExpr (SymExpr sym), IsExprBuilder sym) =>
   sym ->
   (AllocInfo sym -> IO (Pred sym)) ->
   SymNat sym ->
   MemAllocs sym ->
-  IO (Pred sym)
-isAllocatedGeneric sym inAlloc blk = go (pure (falsePred sym))
+  IO (Pred sym, Pred sym)
+isAllocatedGeneric sym inAlloc blk = go (pure (falsePred sym)) (pure (truePred sym))
   where
-    go :: IO (Pred sym) -> MemAllocs sym -> IO (Pred sym)
-    go fallback (MemAllocs []) = fallback
-    go fallback (MemAllocs (alloc : r)) =
+    go :: IO (Pred sym) -> IO (Pred sym) -> MemAllocs sym -> IO (Pred sym, Pred sym)
+    go fallback fallbackFreed (MemAllocs []) = (,) <$> fallback <*> fallbackFreed
+    go fallback fallbackFreed (MemAllocs (alloc : r)) =
       case alloc of
         Allocations am ->
           case asNat blk of
             Just b -> -- concrete block number, look up entry
               case Map.lookup b am of
-                Nothing -> go fallback (MemAllocs r)
-                Just ba -> inAlloc ba
+                Nothing -> go fallback fallbackFreed (MemAllocs r)
+                Just ba -> (,) <$> inAlloc ba <*> fallbackFreed
             Nothing -> -- symbolic block number, need to check all entries
-              Map.foldrWithKey checkEntry (go fallback (MemAllocs r)) am
+              Map.foldrWithKey checkEntry (go fallback fallbackFreed (MemAllocs r)) am
               where
                 checkEntry a ba k =
-                  do sameBlock <- natEq sym blk =<< natLit sym a
-                     itePredM sym sameBlock (inAlloc ba) k
```

Comment:
> This refactoring forces the continuation `k` more often than the original

In this case, we know that `blk` isn't a concrete nat (see `asNat blk` above), so is there any way that `natEq sym blk =<< natLit sym a` could be concrete?

Thanks for the feedback on this point, this is exactly the sort of more subtle performance issue I'm hoping *not* to introduce.

Reply:
FWIW, it looks to me like if we match on `asConstantPred sameBlock` here, we could just return `k` if that's `Just False` but we'd still have to force `k` in the `Just True` case, which is more often than before. :disappointed:

Reply:
Even if `blk` isn't concrete, we still can sometimes concretely evaluate the equality test, particularly if we have nontrivial range information about `blk` in the abstract domain.  For example, if we constructed the value symbolic block value `if b then 5 else 6`, then range information would tell us the value is in the range `[5 .. 6]`, and would let us conclude the the value is not equal to `7`.  We sometimes rewrite equalities under if/then/else also, and that might apply here.

Reply:
Is that because this ties the computation of fallback and fallbackFreed together?

Reply:
Well, let's think this through a bit more.  If `sameBlock` is true here, doesn't that mean that to could not have previously been freed and we can just return `false` in that case?

Reply:
> Even if blk isn't concrete, we still can sometimes concretely evaluate the equality test, particularly if we have nontrivial range information about blk in the abstract domain.

Makes sense to me, thanks! I was forgetting about the abstract domain.

> Is that because this ties the computation of fallback and fallbackFreed together?

That's right.

>  If sameBlock is true here, doesn't that mean that to could not have previously been freed and we can just return false in that case?

That makes sense to me - it couldn't have previously been freed, because we've found where it was allocated and it couldn't have been freed before it's allocated. 

OTOH, while the above example of the abstract domain makes it clear why the equality test could concretely *fail*, even though `blk` isn't concrete, it's not obvious to me how the equality test could concretely *succeed*, unless there's a weird situation like `asNat` "not knowing as much as" `asConstantPred`. That being said, it seems sound to return `(truePred sym, truePred sym)` in the `Just True` case, even if it's unreachable.

Reply:
I've pushed a version with this fixed so that `k` only gets forced when `asConstantPred sameBlock` returns `Nothing`. I added a note about my uncertainty that the `Just True` branch is reachable, which I'm just as happy to leave off.

Reply:
I think there are certain very specific cases where the `Just True` branch might be reachable. The example I have in mind is
```
blk = if p then 5 else (if not p then 5 else 20)
```

If we evaluate concretely if `blk == 5` then we can push the equality down inside the if/then/else, which would give

```
if p then true else (if not p then true else false)
```

and the if/then/else would be able to collapse to `p or (not p)`, which is a specific kind of resolution we can recognize as `true`.

Regardless, I'm OK with the comment because: as you say, the result should be sound either way.

Reply:
> I think there are certain very specific cases where the Just True branch might be reachable.

Thanks, this is another helpful example in building up my understanding of how the abstract interpretation/symbolic execution/rewriting combo works!